index.js

```diff
@@ -243,39 +244,57 @@ export default class Supercluster {
-            let wx = p.x * numPoints;
-            let wy = p.y * numPoints;
-
-            let clusterProperties = reduce && numPoints > 1 ? this._map(p, true) : null;
-
-            // encode both zoom and point index on which the cluster originated -- offset by total length of features
-            const id = (i << 5) + (zoom + 1) + this.points.length;
+            const numPointsOrigin = p.numPoints || 1;
+            let numPoints = numPointsOrigin;
 
+            // count the number of points in a potential cluster
             for (const neighborId of neighborIds) {
                 const b = tree.points[neighborId];
                 // filter out neighbors that are already processed
-                if (b.zoom <= zoom) continue;
-                b.zoom = zoom; // save the zoom (so it doesn't get processed twice)
+                if (b.zoom > zoom) numPoints += b.numPoints || 1;
+            }
+
+            if (numPoints >= minPoints) { // enough points to form a cluster
+                let wx = p.x * numPointsOrigin;
+                let wy = p.y * numPointsOrigin;
+
+                let clusterProperties = reduce && numPointsOrigin > 1 ? this._map(p, true) : null;
+
+                // encode both zoom and point index on which the cluster originated -- offset by total length of features
+                const id = (i << 5) + (zoom + 1) + this.points.length;
+
+                for (const neighborId of neighborIds) {
+                    const b = tree.points[neighborId];
+
+                    if (b.zoom <= zoom) continue;
+                    b.zoom = zoom; // save the zoom (so it doesn't get processed twice)
 
-                const numPoints2 = b.numPoints || 1;
-                wx += b.x * numPoints2; // accumulate coordinates for calculating weighted center
-                wy += b.y * numPoints2;
+                    const numPoints2 = b.numPoints || 1;
+                    wx += b.x * numPoints2; // accumulate coordinates for calculating weighted center
+                    wy += b.y * numPoints2;
 
-                numPoints += numPoints2;
-                b.parentId = id;
+                    b.parentId = id;
 
-                if (reduce) {
-                    if (!clusterProperties) clusterProperties = this._map(p, true);
-                    reduce(clusterProperties, this._map(b));
+                    if (reduce) {
+                        if (!clusterProperties) clusterProperties = this._map(p, true);
+                        reduce(clusterProperties, this._map(b));
+                    }
                 }
-            }
 
-            if (numPoints === 1) {
-                clusters.push(p);
-            } else {
                 p.parentId = id;
                 clusters.push(createCluster(wx / numPoints, wy / numPoints, id, numPoints, clusterProperties));
+
+            } else { // left points as unclustered
+                clusters.push(p);
+
+                if (numPoints > 1) {
```

Comment:
I am curious why we compare numPoints with 1 instead of numPointsOrigin?

Reply:
When `numPointsOrigin` is bigger than `1`, it means that the origin is itself a cluster, which could only be formed from `minPoints` points or more. So in case of origin cluster, we never reach this `else` case (`numPoints` will always be `>= minPoints`).

Reply:
I see! Thank you!